plutus-scb/src/Cardano/Metadata/Client.hs

```diff
@@ -1,40 +1,59 @@
-{-# LANGUAGE GADTs            #-}
-{-# LANGUAGE LambdaCase       #-}
-{-# LANGUAGE RankNTypes       #-}
-{-# LANGUAGE TypeApplications #-}
-{-# LANGUAGE TypeOperators    #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeOperators       #-}
 
-module Cardano.Metadata.Client where
+module Cardano.Metadata.Client
+    ( handleMetadataClient
+    ) where
 
 import           Cardano.Metadata.API      (API)
-import           Cardano.Metadata.Types    (MetadataEffect (GetProperties, GetProperty), Property, PropertyKey, Subject)
+import           Cardano.Metadata.Types    (JSONEncoding (ExternalEncoding),
+                                            MetadataEffect (GetProperties, GetProperty),
+                                            MetadataError (MetadataClientError))
 import           Control.Monad.Freer       (Eff, LastMember, Member, type (~>), interpret, sendM)
 import           Control.Monad.Freer.Error (Error, throwError)
 import           Control.Monad.IO.Class    (MonadIO, liftIO)
+import           Data.Coerce               (coerce)
 import           Data.Proxy                (Proxy (Proxy))
-import           Servant.Client            (ClientEnv, ClientError, ClientM, client, runClientM)
+import           Network.HTTP.Types        (status404)
+import           Servant.Client            (ClientEnv, ClientError (FailureResponse), ClientM, client,
+                                            responseStatusCode, runClientM)
 import           Servant.Extra             (left, right)
 
-getProperties :: Subject -> ClientM [Property]
-getProperty :: Subject -> PropertyKey -> ClientM Property
-(getProperties, getProperty) = (_getProperties, _getProperty)
-  where
-    _getProperties = left . api
-    _getProperty = right . api
-    api = client (Proxy @API)
-
 handleMetadataClient ::
        forall m effs.
-       (LastMember m effs, MonadIO m, Member (Error ClientError) effs)
+       ( LastMember m effs
+       , MonadIO m
+       , Member (Error MetadataError) effs
+       )
     => ClientEnv
     -> Eff (MetadataEffect ': effs) ~> Eff effs
 handleMetadataClient clientEnv =
-    let runClient :: forall a. ClientM a -> Eff effs a
-        runClient a =
-            (sendM $ liftIO $ runClientM a clientEnv) >>= either throwError pure
+    let (getProperties, getProperty) = (_getProperties, _getProperty)
+          where
+            _getProperties = left . api
+            _getProperty = right . api
+            -- This can be flipped between AesonEncoding and
+            -- ExternalEncoding depending on which metadata server
+            -- we're using.
+            api = client (Proxy @(API 'ExternalEncoding))
```

Comment:
:thinking: 

Why do we have two again? They're using something weird which we don't want to follow? We need to clarify things with them?

Reply:
Yeah, this is an interesting one.

We really want our JSON encodings to be Aeson-compatible, because then we get the PureScript version for (nearly) free. When the 3rd-party datasource uses a custom encoding, we have a few options:

### Write custom codex for Haskell, and again for PureScript.

The most work. The most fragile. The highest maintenance cost. Avoid if possible.

### Asymmetric Codex

If the data's only traveling in one direction (say 3rd Party -> Haskell -> PureScript) you can get away with giving the Haskell type a custom decoder and and a default aeson encoder. Then Haskell is reading one format and writing the other. The same argument applies if the data's only travelling in the other direction (PS -> H ->3rd.) 

This isn't the case here. The frontend needs to send and receive these metadata types.

### Persuade the 3rd Party to Use Aeson-Style Encoding

In this case it is an option. We can probably get Steve to change to suit us. _But_, pretty soon we're going to face integrating with the wallet and node clients, and I'm guessing they won't change their encodings for our benefit. I figured that right now now, while the team's under _relatively_ low pressure, it was a good time to bite the bullet and come up with a universal recipe for solving this problem.

### Double the DataTypes

Create two datatypes for each datum. One with custom JSON encoders, one with Aeson defaults. This reduces the problem to maintaining an `Iso` between them, at the cost of having two datatypes for each concept. It doubles the amount of code, but it's all type-checked. And it doesn't involve an "fancy" Haskell. But it feels like a lie to me - we don't have two datatypes, just two encodings. The correct solution would follow that fact.

### Switch Codex Based On A Phantom Type

Add a phantom `encoding` type parameter, and use that to choose between two different `ToJSON/FromJSON` instances. The downside is it's "fancy" Haskell (but not that fancy by our team's standards.) The upside is there's no type duplication, and the solution follows the problem. Using it, you can easily set up a servant API with `Proxy @(API AesonEncoding)` or `Proxy @(API ExternalEncoding)` and the JSON-handling flips over to follow it. And since it's just a phantom type parameter, you can use `coerce` to flip between the datatypes in a type-safe way, with no runtime cost.

This is what I went with. I had a few problems getting it to sit nicely with Freer, but nothing too bad. And now we can just roll out the same recipe for the other servers' encodings as they come up.

Reply:
🏅 most well thought out PR comment ever

Reply:
Great explanation. Now put it in a note in the code :)

Reply:
Commented. 👍